File: src/m27_ctx.c
/* m27_ctx.c -- the context switch on its own, before anything depends on it.
 *
 * src/sys/ctx.c parks a context by copying its extent of the one engine
 * stack into far memory and copying it back at the same addresses when
 * the context runs again.  Two things have to be true for that to be
 * worth building on, and neither of them is visible from the outside of
 * a working desktop:
 *
 *   THE TURNS HAPPEN IN THE ORDER THEY WERE ASKED FOR.  The runner
 *   records who ran, in sequence, in a plain array the host reads by
 *   symbol.  A switch that silently did nothing, or that came back to
 *   the wrong context, shows up as a different sequence and not as a
 *   crash.
 *
 *   A FRAME SURVIVES A SWITCH FROM UNDERNEATH IT.  A context that
 *   switches at the bottom of a recursion has to find its locals intact
 *   on the way back out, at every level.  So one of the two contexts
 *   recurses, writes a value derived from the depth into a local and
 *   into an array, switches away from the deepest frame, and checks each
 *   local against the array as the recursion unwinds.  That is the whole
 *   claim the design rests on -- the extent goes back to the SAME
 *   addresses, so a pointer into a frame is still the address it was --
 *   and a copy that were off by one byte, or short, would fail it.
 *
 * The runner also leaves the measured park sizes where the gate can
 * print them: what a switch actually costs is a number this project
 * would rather measure than argue about.
 */
#include <stdint.h>
#include "sys/ctx.h"
#include "sys/farmem.h"

#define STATUS ((volatile unsigned char *) 0x0600)

/* The call gate's own words.  A context carries them (src/sys/ctx.c) and
 * src/sys/abi.s reaches for them, so linking app_run -- which is how a
 * context enters its program -- means defining them.  The engine behind
 * them is not here and does not need to be: this runner executes no COP,
 * and gem_entry is the proof that it does not.  Reaching it is a failure
 * the gate can see, not a crash. */
uint32_t gem_pb;
uint8_t  gem_which;
uint16_t gem_api_sp;
uint8_t  gem_depth;
uint16_t gem_calls;
uint16_t gem_bad;
uint16_t gem_reached;

void gem_entry(void)
{
    gem_reached++;
}

#define ROUNDS  5           /* turns each of the two contexts takes */
#define DEEP    6           /* frames the recursing context builds */

/* Who ran, in order: 1 and 2 are the two contexts, 0 is the root. */
uint16_t seq[32];
uint16_t nseq;

/* The recursion's evidence: what each level wrote, and what it still
 * held after the switch. */
uint16_t deep_put[DEEP];
uint16_t deep_got[DEEP];
uint16_t deep_bad;          /* levels whose local did not survive */

uint16_t ctx_done;          /* the runner reached the end */
uint16_t rounds_a, rounds_b;

CTX c_root, c_a, c_b;

static void note(uint16_t who)
{
    if (nseq < 32)
        seq[nseq++] = who;
}

/* Down DEEP frames, switch away from the bottom, and check every level
 * on the way back up.  The local is deliberately kept live across the
 * call so the compiler cannot spill it somewhere that is not the stack. */
static void recurse(uint16_t level)
{
    uint16_t mine = (uint16_t)(0xC000 + level * 7);

    deep_put[level] = mine;
    if (level + 1 < DEEP)
        recurse((uint16_t)(level + 1));
    else
        ctx_switch(&c_b);           /* the deepest frame yields */
    deep_got[level] = mine;
    if (mine != (uint16_t)(0xC000 + level * 7))
        deep_bad++;
}

static int16_t proc_a(void)
{
    uint16_t i;

    for (i = 0; i < ROUNDS; i++) {
        note(1);
        rounds_a++;
        ctx_switch(&c_b);
    }
    recurse(0);
    note(0x11);
    return 0;                       /* done: ctx_boot hands back to the root */
}

static int16_t proc_b(void)
{
    for (;;) {
        note(2);
        rounds_b++;
        ctx_switch(&c_a);
    }
}

__task void main(void)
{
    STATUS[0] = 'C';
    STATUS[1] = 'X';
    STATUS[2] = 0;

    farmem_probe();
    if (farmem.kind == FARMEM_NONE) {
        STATUS[2] = 0xFF;
        for (;;)
            ;
    }

    ctx_init(&c_root);
    if (!ctx_make(&c_a, (uint32_t)proc_a) ||
        !ctx_make(&c_b, (uint32_t)proc_b)) {
        STATUS[2] = 0xFE;
        for (;;)
            ;
    }

    note(0);
    ctx_switch(&c_a);               /* comes back when proc_a returns */
    note(0);
    ctx_done = 1;
    STATUS[2] = 1;
    for (;;)
        ;
}

File: tests/emu/m27_ctx.py
#!/usr/bin/env python3
"""Phase 36 gate: two contexts taking turns on one stack.

A desk accessory is a second program resident beside the first, and GEM's
answer to how it gets the processor is a process with its own stack that
blocks inside evnt_multi while the dispatcher runs somebody else.  gem4xe
cannot have a second engine stack -- 1,239 of the 2,048 bytes are used at
the low-water mark and LoRAM has 47 free -- so the contexts SHARE the one
stack: a parked context's extent of it lives in far memory and goes back
to THE SAME ADDRESSES when it runs again.  src/sys/ctx.h has the argument.

Everything downstream of this rests on the copy being exact, so this gate
runs before there is a desktop in the way:

  THE TURNS ARE IN THE ORDER THEY WERE ASKED FOR.  The runner writes who
  ran into a plain array; a switch that quietly did nothing, or came back
  to the wrong context, is a different array and not a crash.

  A FRAME SURVIVES A SWITCH FROM UNDERNEATH IT.  One context recurses six
  frames deep, writes a value derived from the depth into a local, yields
  from the deepest frame, and checks every local on the way back out.  A
  copy short by one byte fails this; nothing else here would notice.

The gate also prints what a park actually COST -- the high-water extent
each context copied -- because "a switch is cheap if it is shallow" is
the claim the design was chosen on and it should be a measurement.
"""
import os
import sys

ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..")
sys.path.insert(0, os.path.join(ROOT, "tools"))
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from a8test.launcher import launch          # noqa: E402
import symfile                              # noqa: E402

DISK = os.path.abspath(os.path.join(ROOT, "build", "m27-boot.atr"))
SYMS = os.path.join(ROOT, "build", "m27.sym")
STATUS = 0x0600

ROUNDS, DEEP = 5, 6
# src/m27_ctx.c: the root, then five turns each, then the deep yield, the
# recursing context's mark and the root again.
WANT_SEQ = [0] + [1, 2] * ROUNDS + [2, 0x11, 0]

# src/sys/ctx.h, in order.  The gate does not take these on trust: the
# three CTX records are consecutive in bss, so their spacing is the
# struct's real size and is checked against the last field's end.
CTX_SP, CTX_SAVE, CTX_LEN, CTX_DEEP, CTX_ENTRY = 0, 2, 6, 8, 10
CTX_LIVE, CTX_PB, CTX_API_SP, CTX_WHICH, CTX_DEPTH = 14, 16, 20, 22, 24
CTX_SIZE = 26


def word(b, addr):
    d = bytes(b.memdump(addr, 2))
    return d[0] | (d[1] << 8)


def words(b, addr, n):
    d = bytes(b.memdump(addr, 2 * n))
    return [d[2 * i] | (d[2 * i + 1] << 8) for i in range(n)]


def main(argv):
    fails = []

    def check(cond, msg):
        if not cond:
            fails.append(msg)
            print(f"  FAIL: {msg}")
        return cond

    syms = symfile.load(SYMS)
    emu = launch(tag="m27", memsize="1088K", extra_args=["--disk", DISK])
    b = emu.bridge
    try:
        b.frames(300)
        b.poke(0xD1FF, 0x01)
        b.poke(0xD191, 0x00)                # -> 65C816 (resets; DOS reboots)
        b.frames(500)
        for k in ("M", "2", "7", "RETURN"):
            b.key(k)
            b.frames(6)
        for _ in range(0, 3000, 50):
            b.frames(50)
            if bytes(b.memdump(STATUS, 3)) == b"CX\x01":
                break
        st = bytes(b.memdump(STATUS, 3))
        if not check(st == b"CX\x01",
                     f"the runner did not finish (status {st!r})"):
            return 1
        cpu = b.cmd("HWSTATE")["cpu"]["mode"]
        check(cpu == "65C816", f"CPU is {cpu}")

        # The record's size, from the linker rather than from this file.
        c_root, c_a, c_b = syms["c_root"], syms["c_a"], syms["c_b"]
        span = sorted((c_root, c_a, c_b))
        check(span[1] - span[0] == CTX_SIZE and span[2] - span[1] == CTX_SIZE,
              f"CTX is not {CTX_SIZE} bytes: the records are at "
              f"{[hex(x) for x in span]}")

        # 1. the order of the turns
        nseq = word(b, syms["nseq"])
        seq = words(b, syms["seq"], min(nseq, 32))
        check(seq == WANT_SEQ,
              f"the turns ran {seq}, expected {WANT_SEQ}")
        print(f"  {nseq} turns, in order: "
              f"{' '.join('root' if s == 0 else 'A' if s == 1 else 'B'
                          if s == 2 else 'A-done' for s in seq)}")
        check(word(b, syms["rounds_a"]) == ROUNDS,
              f"context A took {word(b, syms['rounds_a'])} turns, not {ROUNDS}")
        check(word(b, syms["rounds_b"]) == ROUNDS + 1,
              f"context B took {word(b, syms['rounds_b'])} turns, "
              f"not {ROUNDS + 1}")

        # 2. the recursion's locals, level by level
        put = words(b, syms["deep_put"], DEEP)
        got = words(b, syms["deep_got"], DEEP)
        bad = word(b, syms["deep_bad"])
        check(put == got and bad == 0,
              f"{bad} frame(s) did not survive the switch: wrote {put}, "
              f"found {got}")
        print(f"  {DEEP} frames deep, every local intact across the yield "
              f"({put[0]:#06x}..{put[-1]:#06x})")

        # 3. nothing went wrong quietly
        check(word(b, syms["ctx_over"]) == 0,
              f"{word(b, syms['ctx_over'])} park(s) refused for want of room")
        check(word(b, syms["gem_reached"]) == 0,
              "a COP reached the gate's stub: the runner is not supposed "
              "to make one")

        # 4. what it cost
        for name, addr in (("root", c_root), ("A", c_a), ("B", c_b)):
            deep = word(b, addr + CTX_DEEP)
            live = word(b, addr + CTX_LIVE)
            print(f"  context {name}: {deep} bytes at its deepest park, "
                  f"live={live}")
        check(word(b, c_a + CTX_LIVE) == 2,
              "context A did not end marked done")
    finally:
        emu.stop()

    print(f"gem4xe-m27: {'PASS' if not fails else 'FAIL'} -- two contexts, "
          f"one stack, {len(fails)} problem(s)")
    return 1 if fails else 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
